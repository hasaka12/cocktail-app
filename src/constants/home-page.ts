export type HomeKey = "Gin" | "Vodka" | "Wine" | "Tequila";

interface HomeInfo {
  key: HomeKey;
  name: string;
}

export const homePageData: Record<string, HomeInfo> = {
  gin: { key: "Gin", name: "Gin Cocktails" },
  vodka: { key: "Vodka", name: "Vodka Cocktails" },
  Wine: { key: "Wine", name: "Wine Cocktails" },
  Tequila: { key: "Tequila", name: "Tequila Cocktails" },
};
