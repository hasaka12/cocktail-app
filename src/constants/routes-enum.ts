export enum Route {
  Home = "home",
  Search = "search",
  Favorite = "favorite",
  All = "all",
  Settings = "settings",
}
