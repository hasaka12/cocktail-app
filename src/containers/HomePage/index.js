import React, { useEffect, useState } from "react";
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import Container from "@mui/system/Container";
import Button from '@mui/material/Button';
import Refresh from '@mui/icons-material/Refresh';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import CircularProgress from '@mui/material/CircularProgress';

import { randomCocktails } from "../../actions/postActions";
import FavoriteCocktails from "../FavoriteCocktails";
import SearchBar from "../SearchBar";
import ActionAreaCard from "../../components/ActionAreaCard";

const HomePage = props => {
    const [randomList, setRandomList] = useState(null);

    useEffect(() => {
        props.randomCocktails();
    }, [])

    useEffect(() => {
        if (props.randomCocktailList && props.randomCocktailList.length > 0) {
            setRandomList(props.randomCocktailList);
        }
    }, [props.randomCocktailList])

    const onRefresh = () => {
        setRandomList(null);
        props.randomCocktails();
    }

    return (
        <Container>
            <SearchBar />
            <FavoriteCocktails />
            <h3>Random Cocktails</h3>
            <Button 
                variant="outlined" 
                startIcon={<Refresh />} 
                onClick={onRefresh}
                sx={{ mb: 2 }}
            >
                Refresh
            </Button>
            <Box sx={{ flexGrow: 1 }}>
                <Grid container sx={{ pl: 2 }}>
                    {randomList && randomList.length > 0 ? randomList.map((data, i) => (
                        <Grid key={i} sx={{ m: 2 }} item xs>
                            <ActionAreaCard 
                                title={data.drinks[0].strDrink && data.drinks[0].strDrink}
                                desc={data.drinks[0].strCategory && data.drinks[0].strCategory}
                                img={data.drinks[0].strDrinkThumb && data.drinks[0].strDrinkThumb}
                            />
                        </Grid>
                    )): <CircularProgress />}
                </Grid>
            </Box>
        </Container>
    );

}

const randomCocktailFunc = dispatch => {
    return bindActionCreators({ randomCocktails }, dispatch)
}

const mapStateToProps = state => {
    return {
        randomCocktailList: state.randomCocktails
    }
}
  
export default connect(mapStateToProps, randomCocktailFunc)(HomePage);