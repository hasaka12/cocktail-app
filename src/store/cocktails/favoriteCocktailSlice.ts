import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { CocktailsDto } from "../../models";

interface CounterState {
  data: CocktailsDto[];
}

const initialState: CounterState = {
  data: [],
};

const counterSlice = createSlice({
  name: "my-counter",
  initialState,
  reducers: {
    addFavoriteCocktails: (state, action: PayloadAction<CocktailsDto>) => {
      if (state.data.some((item) => item.idDrink === action.payload.idDrink)) {
        const index = state.data.findIndex(
          (item) => item.idDrink === action.payload.idDrink
        );
        state.data.splice(index, 1);
        return;
      }
      state.data.push(action.payload);
    },
  },
});

export const { addFavoriteCocktails } = counterSlice.actions;

export default counterSlice.reducer;
