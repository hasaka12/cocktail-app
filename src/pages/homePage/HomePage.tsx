import { Fragment } from "react";

import Typography from "@mui/material/Typography";
import Divider from "@mui/material/Divider";

import { homePageData } from "../../constants/home-page";

import HomeCard from "./HomeCard";

export default function HomePage() {
  return (
    <div className="mx-4">
      {Object.values(homePageData).map((category) => (
        <Fragment key={category.key}>
          <div className="my-10">
            <Divider>
              <Typography variant="h4" component="h4">
                {category.name}
              </Typography>
            </Divider>
          </div>
          <HomeCard keyValue={category.key} />
        </Fragment>
      ))}
    </div>
  );
}
