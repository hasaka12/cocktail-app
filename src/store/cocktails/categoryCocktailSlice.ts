import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import axios from "axios";

import { CategoryCocktailDto } from "../../models/CategoryCocktailDto";
import { HomeKey } from "../../constants/home-page";
import { BASE_URL } from "../../constants/app";

interface CounterState {
  isLoading: boolean;
  data: CategoryCocktailDto[];
}

const initialState: CounterState = {
  isLoading: false,
  data: [],
};

const categorySearchCocktailSlice = createSlice({
  name: "categorySearchCocktail",
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getCategorySearchCocktail.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(getCategorySearchCocktail.fulfilled, (state, action) => {
        state.isLoading = false;
        state.data.push(action.payload);
      });
  },
});

export const getCategorySearchCocktail = createAsyncThunk(
  "getCategorySearchCocktail",
  async (key: HomeKey) => {
    const url = `${BASE_URL}/filter.php?i=${key}`;
    try {
      const res = await axios.get(url);
      return { ...res.data, key };
    } catch (err) {
      console.error(err);
      return null;
    }
  }
);

export default categorySearchCocktailSlice.reducer;
