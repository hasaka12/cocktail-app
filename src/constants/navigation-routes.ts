import { Route } from "./routes-enum";

export const RoutesMap = {
  [Route.Home]: () => "/",
  [Route.Search]: () => "/search",
  [Route.Favorite]: () => "/favorite",
  [Route.All]: () => "/all",
  [Route.Settings]: () => "/settings",
};

export type RouteKeys = keyof typeof RoutesMap;

// utility types to help us generate `RouteKeysWithParams` and `RouteKeysWithoutParams`
type IsRouteMapHasParams<TRoute extends RouteKeys> = Parameters<
  (typeof RoutesMap)[TRoute]
> extends []
  ? never
  : TRoute;
type IsRouteMapHasNoParams<TRoute extends RouteKeys> = Parameters<
  (typeof RoutesMap)[TRoute]
> extends []
  ? TRoute
  : never;

// wrapper functions to distribute generic type
type IsRouteMapHasParamsWrapper<TRoute extends RouteKeys> =
  TRoute extends RouteKeys ? IsRouteMapHasParams<TRoute> : never;
type IsRouteMapHasNoParamsWrapper<TRoute extends RouteKeys> =
  TRoute extends RouteKeys ? IsRouteMapHasNoParams<TRoute> : never;
// end of utility types

export type RouteKeysWithParams = IsRouteMapHasParamsWrapper<RouteKeys>;
export type RouteKeysWithoutParams = IsRouteMapHasNoParamsWrapper<RouteKeys>;

export type RouteParams<TRoute extends RouteKeys> = Parameters<
  (typeof RoutesMap)[TRoute]
> extends Array<infer TParams>
  ? TParams
  : never;
