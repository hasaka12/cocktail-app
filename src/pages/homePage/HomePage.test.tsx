import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { Store, UnknownAction, configureStore } from "@reduxjs/toolkit";

import HomePage from "./HomePage";
import categoryCocktailSlice from "../../store/cocktails/categoryCocktailSlice";
import favoriteCocktailSlice from "../../store/cocktails/favoriteCocktailSlice";
import { homePageData } from "../../constants/home-page";

describe("HomePage", () => {
  let store: Store<unknown, UnknownAction, unknown>;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        favoriteCocktail: favoriteCocktailSlice,
        categoryCocktail: categoryCocktailSlice,
      },
    });
  });

  test("renders HomeCard for each category", () => {
    const { getByText } = render(
      <Provider store={store}>
        <HomePage />
      </Provider>
    );

    Object.values(homePageData).forEach((category) => {
      expect(getByText(category.name)).toBeInTheDocument();
    });
  });

  test("Renders correctly", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <HomePage />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
