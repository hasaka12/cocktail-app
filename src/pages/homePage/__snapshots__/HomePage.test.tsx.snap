// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Renders correctly 1`] = `
<DocumentFragment>
  <div
    class="mx-4"
  >
    <div
      class="my-10"
    >
      <div
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1puturl-MuiDivider-root"
        role="separator"
      >
        <span
          class="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Gin Cocktails
          </h4>
        </span>
      </div>
    </div>
    <div
      class="lg:grid lg:grid-cols-5 lg:gap-4"
    >
      <div
        class="flex m-2 items-center justify-center"
      >
        <div
          class="flex justify-center bouncing-loader gap-2"
        >
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 0ms;"
          />
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 200ms;"
          />
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 500ms;"
          />
        </div>
      </div>
    </div>
    <div
      class="my-10"
    >
      <div
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1puturl-MuiDivider-root"
        role="separator"
      >
        <span
          class="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Vodka Cocktails
          </h4>
        </span>
      </div>
    </div>
    <div
      class="lg:grid lg:grid-cols-5 lg:gap-4"
    >
      <div
        class="flex m-2 items-center justify-center"
      >
        <div
          class="flex justify-center bouncing-loader gap-2"
        >
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 0ms;"
          />
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 200ms;"
          />
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 500ms;"
          />
        </div>
      </div>
    </div>
    <div
      class="my-10"
    >
      <div
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1puturl-MuiDivider-root"
        role="separator"
      >
        <span
          class="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Wine Cocktails
          </h4>
        </span>
      </div>
    </div>
    <div
      class="lg:grid lg:grid-cols-5 lg:gap-4"
    >
      <div
        class="flex m-2 items-center justify-center"
      >
        <div
          class="flex justify-center bouncing-loader gap-2"
        >
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 0ms;"
          />
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 200ms;"
          />
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 500ms;"
          />
        </div>
      </div>
    </div>
    <div
      class="my-10"
    >
      <div
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1puturl-MuiDivider-root"
        role="separator"
      >
        <span
          class="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Tequila Cocktails
          </h4>
        </span>
      </div>
    </div>
    <div
      class="lg:grid lg:grid-cols-5 lg:gap-4"
    >
      <div
        class="flex m-2 items-center justify-center"
      >
        <div
          class="flex justify-center bouncing-loader gap-2"
        >
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 0ms;"
          />
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 200ms;"
          />
          <div
            class="bg-blue-500 w-3 h-3 rounded-full animate-bounce"
            style="animation-delay: 500ms;"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
